### PR TITLE
Mysql datastore interface implementation: Find query

### LINKED ISSUE #1638: Support MySQL as Control-plane datastore: Datastore interface implementation
**What would you like to be added**:

Related to #1581 

**Why is this needed**:

### PR DESCRIPTION
**Pull request by khanhtc1202**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Ref #1638

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/datastore/mysql/iterator.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"database/sql"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+)
+
+// Iterator for MySQL result set
+type Iterator struct {
+	cur *sql.Rows
```
Comment (nghialv) [inline pkg/datastore/mysql/iterator.go:25]: nit: IMO, `rows` is more readable than `cur`

**Comment by nghialv** [comment]
Overall, this looks good. Nice.
@khanhtc1202 Can you raise an issue to add integration tests at [here](https://github.com/pipe-cd/pipe/tree/master/test/integration) for this MySQL package in the future?
(Don't have to do that at this time, but it will be good in the future.)

File: pkg/datastore/mysql/iterator.go (hunk 1 above)
Comment (khanhtc1202) [reply pkg/datastore/mysql/iterator.go]: addressed on https://github.com/pipe-cd/pipe/pull/1678/commits/31d25ab8c9f4ab0739e7468c947fb7141c33079e 🙏

**Comment by khanhtc1202** [comment]
> Overall, this looks good. Nice.
> @khanhtc1202 Can you raise an issue to add integration tests at [here](https://github.com/pipe-cd/pipe/tree/master/test/integration) for this MySQL package in the future?
> (Don't have to do that at this time, but it will be good in the future.)

Sure, created at https://github.com/pipe-cd/pipe/issues/1680 🙏

File: pkg/datastore/mysql/query.go [hunk 2]
```diff
@@ -31,3 +36,60 @@ func buildPutQuery(table string) string {
...
+	}
+	return fmt.Sprintf("WHERE %s", strings.Join(conds[:], " AND "))
+}
+
+func buildOrderByClause(orders []datastore.Order) string {
+	if len(orders) == 0 {
+		return ""
+	}
+
+	var conds []string
```
Comment (nakabonne) [inline pkg/datastore/mysql/query.go:68]: nits: Better pre-allocate.
```suggestion
	conds := make([]string, 0, len(orders))
```

File: pkg/datastore/mysql/query.go [hunk 3]
```diff
@@ -31,3 +36,60 @@ func buildPutQuery(table string) string {
...
+	)
+	return strings.Join(strings.Fields(rawQuery), " ")
+}
+
+func buildWhereClause(filters []datastore.ListFilter) string {
+	if len(filters) == 0 {
+		return ""
+	}
+
+	var conds []string
```
Comment (nakabonne) [inline pkg/datastore/mysql/query.go:56]: Here as well:
```suggestion
	conds := make([]string, 0, len(filters))
```

File: pkg/datastore/mysql/mysql.go [hunk 4]
```diff
@@ -79,7 +79,21 @@ func NewMySQL(url, database string, opts ...Option) (*MySQL, error) {
 
 // Find implementation for MySQL
 func (m *MySQL) Find(ctx context.Context, kind string, opts datastore.ListOptions) (datastore.Iterator, error) {
-	return nil, datastore.ErrUnimplemented
+	var filterVals []interface{}
```
Comment (nakabonne) [inline pkg/datastore/mysql/mysql.go:82]:
```suggestion
	filterVals := make([]interface{}, 0, len(opts.Filters))
```

File: pkg/datastore/mysql/query.go (hunk 2 above)
Comment (khanhtc1202) [reply pkg/datastore/mysql/query.go]: You're right 👍 Addressed by https://github.com/pipe-cd/pipe/pull/1678/commits/b0a18ae6cd330aacab382b8331196cdbea7f9223 🙏

**Comment by nghialv** [comment]
Nice. 👍 
/lgtm

**Comment by nakabonne** [comment]
There you go!
/approve

**Pull request merged**